### PR TITLE
fix(broker): shut down the app-server broker after an idle timeout

### PR DESCRIPTION
**Pull request by cjsteigerwald**
## Problem

`app-server-broker.mjs` has no idle timeout and no lifetime cap. It exits on exactly three paths — a `broker/shutdown` RPC, `SIGTERM`, or `SIGINT`. Nothing in the protocol tells a broker that its client is gone for good, so any broker whose client disappears without sending `broker/shutdown` stays resident until the machine reboots, holding its `/tmp/cxc-*` socket dir and ~40MB of RSS.

## This reproduces from the repo's own test suite

This is not a rare edge case — `npm test` causes it on every run.

Each test in `tests/runtime.test.mjs` creates a fresh temp workspace; `ensureBrokerSession` spawns one broker per workspace; nothing tears them down when the test ends. A single `node --test tests/runtime.test.mjs` run leaves **~58 brokers** behind.

Measured on one machine after three suite runs:

```
$ pgrep -fc 'app-server-broker\.mjs'
174

$ ps -eo time,cmd | grep '[a]pp-server-broker' | awk '{print $1}' | sort -u
00:00:00          # every one idle, zero CPU consumed

$ ss -xp | grep -c app-server-broker
0                 # zero established connections
```

All 174 had been reparented (their spawning shells were long gone) and together held ~2.9GB of unique resident memory on a 7.8GB host, which took available memory from 5.3GB down to 2.3GB.

## Fix

The broker now shuts itself down once it has been idle for `--idle-timeout` milliseconds (default **30 minutes**, overridable with `CODEX_COMPANION_BROKER_IDLE_MS`, explicit `0` disables).

Idle means **no connected sockets and nothing in flight** (`sockets.size === 0 && activeRequestSocket === null && activeStreamSocket === null`). Holding an open connection is sufficient to keep the broker alive, so a long streaming turn can never be cut off mid-flight.

### Why the broker decides, rather than an external sweeper

The broker is the only party that can observe whether it is currently serving anyone. Scanning a job registry from outside and then killing is racy by construction — a job can start between the scan and the kill — and on a machine running several clients concurrently, the broker processes and the `/tmp/cxc-*` namespace are shared, so a pattern-based sweep cannot distinguish one client's orphan from another's live runtime. Putting the decision inside the process removes the race entirely.

### Why it does not clear `broker.json`

An idle exit leaves the `broker.json` session record behind. That is already handled: `ensureBrokerSession` probes the recorded endpoint via `isBrokerEndpointReady` and, when it does not answer, calls `teardownBrokerSession` + `clearBrokerSession` and spawns a replacement. The cost is a single 150ms probe on next use.

The broker deliberately does **not** clear that record itself, even though it knows its own `--cwd`: a replacement broker may already have been spawned for the same cwd and rewritten the record, and clearing it would delete the live broker's endpoint.

### Input handling

Blank and whitespace-only values fall back to the default rather than parsing as `0`, because `Number("  ") === 0` would otherwise silently mean "never expire" — the exact failure this change exists to prevent. Non-numeric and negative values are rejected loudly instead of quietly defaulting.

## Known limitation

There is a narrow residual race: a client can connect between the timer's idle re-check and `shutdown()` completing. It is benign in practice, because a client that finds a dead endpoint goes through the existing teardown-and-respawn path rather than failing, but it is narrowed rather than eliminated, and I would rather state that than claim otherwise.

## Tests

Adds `tests/broker-idle.test.mjs` (5 tests), all spawning a real broker as a subprocess against the existing fake-codex fixture:

- exits on its own after the idle timeout, and `shutdown()` still removes the socket and pidfile
- stays alive while a client is connected, then exits after it disconnects
- `--idle-timeout 0` never idles out
- a non-numeric `--idle-timeout` is rejected rather than silently never expiring
- a blank `--idle-timeout` falls back to the default rather than disabling

Verified non-vacuous by running the new tests against the unpatched tree at `db52e28`: 3 of the 5 fail without the change. The `--idle-timeout 0` case passes both ways by design — it is a guard against idle shutdown becoming unconditional, not coverage of the feature.

Full suite: **100 passing / 0 failing of 100.**

_Correction: an earlier revision of this description reported 3 pre-existing failures on `main`. That was wrong. They were caused by `CODEX_COMPANION_SESSION_ID` being set in my shell: `filterJobsForCurrentClaudeSession` then filters jobs to `job.sessionId === sessionId`, and the `status`/`result` test fixtures are handcrafted without a `sessionId`, so they were all filtered out. With that variable unset the suite is green. There are no pre-existing failures._

## Compatibility

Default-on with a 30 minute timeout. Because `ensureBrokerSession` already respawns on a dead endpoint, the worst case for a client that returns after 30 idle minutes is one cold broker start, not a failure. Set `CODEX_COMPANION_BROKER_IDLE_MS=0` to restore the previous always-resident behaviour.

### CODE REVIEW

File: plugins/codex/scripts/app-server-broker.mjs [hunk 1]
```diff
@@ -63,13 +92,50 @@ async function main() {
...
+      idleTimer = null;
+      // Re-check at fire time: a client may have connected while the timer was pending.
+      if (!isIdle()) {
+        armIdleTimer();
+        return;
+      }
+      void shutdown(server).then(
+        () => process.exit(0),
+        () => process.exit(0)
+      );
```
[the remark this answers is not in the document]
Comment (cjsteigerwald) [reply plugins/codex/scripts/app-server-broker.mjs:161]: Good catch — fixed in 09338b1.

`shutdown()` now closes the listening server before the app-server teardown:

```js
const closed = new Promise((resolve) => server.close(resolve));
for (const socket of sockets) {
  socket.end();
}
await closed;
await appClient.close().catch(() => {});
```

`server.close()` stops listening immediately and resolves once existing connections drain, so the endpoint is no longer accepting during the `appClient.close()` await — which removes both failure modes you describe: a readiness probe selecting a shutting-down broker, and an accepted socket leaving `server.close()` waiting on an already-closed app server.

Since `shutdown()` is shared, the same reordering also covers the SIGTERM and SIGINT paths. The `--idle-timeout 0` test now asserts socket and pidfile cleanup after SIGTERM so that path is covered.

File: plugins/codex/scripts/app-server-broker.mjs [hunk 2]
```diff
@@ -11,6 +11,33 @@ import { parseBrokerEndpoint } from "./lib/broker-endpoint.mjs";
...
+  // value would otherwise DISABLE idle shutdown silently. Explicit "0" is the only
+  // way to turn it off; anything blank falls back to the default.
+  const text = raw === undefined || raw === null ? "" : String(raw).trim();
+  if (text === "") {
+    return DEFAULT_IDLE_TIMEOUT_MS;
+  }
+  const parsed = Number(text);
+  // Reject rather than silently falling back: a typo'd timeout that quietly became
+  // "never expire" would reintroduce the exact leak this exists to close.
+  if (!Number.isFinite(parsed) || parsed < 0) {
```
[the remark this answers is not in the document]
Comment (cjsteigerwald) [reply plugins/codex/scripts/app-server-broker.mjs:37]: Agreed — fixed in 09338b1.

Values above `2147483647` ms are now rejected at startup rather than accepted:

```
Invalid idle timeout "2592000000": must be at most 2147483647 ms (Node timer limit).
```

Rejecting rather than capping, for consistency with how non-numeric and negative values are already handled here — silently substituting a different duration than the one requested is the same class of surprise as the overflow itself.

Two tests cover it: a ~30-day timeout is rejected, and exactly `2147483647` is accepted so the guard is not off by one.

File: plugins/codex/scripts/app-server-broker.mjs [hunk 3]
```diff
@@ -63,13 +101,50 @@ async function main() {
...
+    disarmIdleTimer();
+    if (idleTimeoutMs <= 0 || !isIdle()) {
+      return;
+    }
+    idleTimer = setTimeout(() => {
+      idleTimer = null;
+      // Re-check at fire time: a client may have connected while the timer was pending.
+      if (!isIdle()) {
+        armIdleTimer();
+        return;
```
[the remark this answers is not in the document]
Comment (cjsteigerwald) [reply plugins/codex/scripts/app-server-broker.mjs:156]: Confirmed and fixed in de754b0 — this was a real hole, thanks.

I traced the interleaving you describe: `activeRequestSocket = socket` is set before the await, the close handler clears ownership and arms the timer, the response then assigns the already-closed socket to `activeStreamSocket`, the pending timer sees a "busy" broker and declines to re-arm, and `routeNotification` clears the stale ownership without scheduling.

Fixed in two places, because there is a second variant:

1. `armIdleTimer()` is now called when ownership is released — in `routeNotification` on `turn/completed`, and on both the success and error paths of a request. As you say, the owning socket may already be gone, so no further close handler will fire.

2. `isIdle()` no longer treats ownership held by a *destroyed* socket as busy. That covers the case where `turn/completed` never arrives at all, which would otherwise pin the broker open indefinitely. A destroyed socket has nobody listening, so there is no stream left to protect.

Added a test that starts a thread, fires `turn/start`, and destroys the socket without reading the response. Against 09338b1 it fails exactly as predicted:

```
not ok 8 - broker still idles out after a client abandons a streaming request
  error: broker stayed resident after the client abandoned a stream
```

Full suite: 96 passing / 3 failing of 99, the 3 pre-existing on main and unrelated to this PR.

**Comment by cjsteigerwald** [comment]
@codex review

All three findings from the earlier passes are addressed, but the latest commit (de754b0) has not been reviewed yet — the previous passes covered e19c574 and 09338b1. Please review the current head.

Summary of what changed since your last pass:

- `armIdleTimer()` is now also called when ownership is released (in `routeNotification` on `turn/completed`, and on both the success and error paths of a request), so an orphaned stream still schedules a shutdown.
- `isIdle()` no longer counts ownership held by a *destroyed* socket as busy, covering the variant where `turn/completed` never arrives at all.

Both are in `plugins/codex/scripts/app-server-broker.mjs`. The concurrency around ownership release and timer scheduling is the part most worth scrutinising.

File: plugins/codex/scripts/app-server-broker.mjs [hunk 4]
```diff
@@ -63,13 +101,58 @@ async function main() {
...
+  // to activeStreamSocket. A destroyed socket has nobody listening, so treating it as
+  // "busy" would pin the broker open forever if turn/completed never arrives.
+  function ownershipIsLive(socket) {
+    return socket !== null && !socket.destroyed;
+  }
+
+  // Idle means nobody is connected AND nothing is in flight. Holding an open socket is
+  // enough to keep the broker alive, so a long streaming turn can never be cut short.
+  function isIdle() {
+    return sockets.size === 0 && !ownershipIsLive(activeRequestSocket) && !ownershipIsLive(activeStreamSocket);
```
[the remark this answers is not in the document]
Comment (cjsteigerwald) [reply plugins/codex/scripts/app-server-broker.mjs]: Confirmed and fixed in cfc4188 — and you are right that this was worse than the original leak, since callers fall back to direct app servers and the broker never recovers.

This was a defect the previous commit introduced: de754b0 taught `isIdle()` to tolerate a destroyed owner, but the busy guard and the notification target still tested for non-null, so the three checks could disagree.

Rather than teach one more predicate to tolerate a stale owner, I release it. `releaseDeadOwnership()` clears `activeRequestSocket`, and clears `activeStreamSocket` together with `activeStreamThreadIds`, whenever the socket is destroyed:

```js
function releaseDeadOwnership() {
  if (activeRequestSocket !== null && activeRequestSocket.destroyed) {
    activeRequestSocket = null;
  }
  if (activeStreamSocket !== null && activeStreamSocket.destroyed) {
    activeStreamSocket = null;
    activeStreamThreadIds = null;
  }
}
```

It runs at all five points where ownership is inspected: arming the timer, the timer firing, a client connecting, before the busy guard, and before choosing a notification target. `isIdle()` returns to plain null checks, so there is no longer a second notion of liveness that can drift from the first — which was the underlying problem rather than any individual check.

Test: a client abandons a streaming request, then a second client connects and issues `thread/start`, which must be served rather than refused. Against de754b0 it fails with the reconnecting client rejected.

Full suite: 97 passing / 3 failing of 100, the 3 pre-existing on main.

**Comment by cjsteigerwald** [comment]
@codex review

cfc4188 is the current head and has not been reviewed yet — the previous passes covered e19c574, 09338b1 and de754b0.

This round is worth a careful look because the last commit introduced the defect you caught in the previous pass. The change replaces the `ownershipIsLive` predicate with `releaseDeadOwnership()`, which clears `activeRequestSocket`, and `activeStreamSocket` together with `activeStreamThreadIds`, whenever the owning socket is destroyed. It is called at five points: arming the idle timer, the timer firing, a client connecting, before the busy guard, and before selecting a notification target.

Specifically worth scrutinising:

- whether clearing ownership before choosing a notification target can drop a notification that a live client still needed
- whether any path still reads `activeRequestSocket` / `activeStreamSocket` without releasing first
- whether releasing on connect can race a request already being dispatched on another socket

**Comment by cjsteigerwald** [comment]
**Correction to my earlier test-suite claims on this PR.**

I reported 3 pre-existing failures on `main` (`status shows phases...`, `status preserves adversarial review kind labels`, `result returns the stored output...`). That was wrong, and the cause is worth knowing because it is a property of the suite rather than of my machine.

`CODEX_COMPANION_SESSION_ID` was set in my shell. When it is set, `filterJobsForCurrentClaudeSession` narrows jobs to `job.sessionId === sessionId`:

```js
function filterJobsForCurrentClaudeSession(jobs) {
  const sessionId = getCurrentClaudeSessionId();
  if (!sessionId) return jobs;
  return jobs.filter((job) => job.sessionId === sessionId);
}
```

The `status` and `result` tests write their fixtures by hand and none carries a `sessionId`, so with that variable set every fixture job is filtered out and the commands correctly report `No jobs recorded yet`. The production behaviour is right; the fixtures simply predate the filter.

Unsetting the variable turns all three green, and every branch is fully green:

```
#677  94 passing / 0 failing of 94
#680  100 passing / 0 failing of 100
#681  93 passing / 0 failing of 93
```

So: **there are no pre-existing failures on `main`** — please disregard that part of my earlier comments.

One thing this does suggest: since this plugin is for Claude Code, contributors are likely to run `npm test` from inside a Claude Code session, where `CODEX_COMPANION_SESSION_ID` is set and these three tests fail for reasons unrelated to their change. Isolating that variable in the test harness (or giving the fixtures a matching `sessionId`) would remove a confusing false negative. Happy to send that as a separate PR if it would be welcome.

**Comment by thossullivan** [comment]
## Independent Verification — macOS, plugin 1.0.6

Installed at commit `cfc4188a22c8704e4a6b20234b776fb074b44b5c`.

### Before cleanup

| Metric | Value |
| --- | --- |
| Broker / app-server trees | 20 |
| Basic Memory wrapper + Python pairs | 39 |
| Resident memory | ~2.2 GiB |

Broker state:

- 15 referenced deleted working directories.
- 4 had no client and no live worker.
- 1 had an active review and was left untouched.

### Verification results

- [X] 9 focused idle tests passed.
- [X] All 100 repository tests passed.
- [X] `npm run build` passed.
- [X]  One real isolated Codex task completed successfully.
- [X]  A 700 ms test timeout reaped its broker; the app-server and MCP child also exited.

### After manual graceful cleanup

| Metric | Before | After |
| --- | --- | --- |
| Brokers | 20 | 1 |
| Basic Memory pairs | 39 | 8 |
| Memory reclaimed | — | ~1.5 GiB |

### Remaining concern

The full suite retained **29 test brokers** through the default 30-minute grace
period. A short test-only timeout would avoid temporary CI and local memory
pressure.

### Conclusion

These results independently confirm both the root cause and the idle-timeout fix.
